cpp/src/utilities/device_atomics.cuh

```diff
@@ -543,13 +469,25 @@ template <typename T, typename BinaryOp>
 __forceinline__  __device__
 T genericAtomicOperation(T* address, T const & update_value, BinaryOp op)
 {
+    // unwrap the input type to expect
```

Comment:
Instead of needing the `cudf::detail::genericAtomicOperationUnderlyingType` function to resolve the underlying type, how about an overload for `genericAtomicOperation` specifically for the `wrapper` types? For example:

```
template <typename T, gdf_dtype dtype, typename BinaryOp, typename W = cudf::detail::wrapper<T, dtype> >
W genericAtomicOperator( W* address, W const& update_valid, BinaryOp op){
     cudf::detail::genericAtomicOperationImpl<T, BinaryOp, sizeof(T)>{}
            (static_cast<T*>(address), cudf::detail::unwrap(update_value), op);
}
```

Reply:
Great pointing!
Now I've removed `cudf::detail::genericAtomicOperationUnderlyingType`.